### DISCUSSION TITLE
Bayesian network: how to execute multiple inferences faster in the same network with different evidences?

### DISCUSSION THREAD

**Discussion by ferencbartok** [pgmpy/pgmpy#1747] [Q&A]
We have this time-consuming calculation in our system for Bayesian networks:

- do initial inference for all nodes/variables with 0 evidence (I use node and variable interchangeably)
- iterate over ~40-50-60% of our nodes (let's call them "observables"; size depends on the given network)
  - for every observable node set one of the states as evidence (do it for all possible states, which number can usually be around 2-3-4)
  - calculate inference for the other 40-50-60% of nodes*
  - calculate a value for the observable node which shows how much effect/impact changing it's state has on the rest of the variables (~comparing original inference with 0 evidence to the just now calculated probabilities) 

This solution is wasteful as most of the things we recalculate should be the same as before. I'm wondering if you could suggest some way to do this calculation faster.?
I know it can be done in parallel, but do you think there are other ways to speed this up? (We are just setting 1 thing as evidence (other than originally existing evidences) and calculating inference for x% of the nodes and doing this many times.)

**Comment by ankurankan** [comment]
Ref #1746
